lib/src/functions.dart

```diff
@@ -61,12 +61,14 @@ Map<T, List<S>> groupBy<S, T>(Iterable<S> values, T Function(S) key) {
 /// The values returned by [orderBy] are compared using the [compare] function.
 /// If [compare] is omitted, values must implement [Comparable<T>] and they are
 /// compared using their [Comparable.compareTo].
-S minBy<S, T>(Iterable<S> values, T Function(S) orderBy,
-    {int Function(T, T) compare}) {
+///
+/// Returns `null` if [values] is empty.
```

Comment:
We should consider whether an exception would be better.

We should also consider if we should drop the top level method in favor of an extension.

Reply:
I would prefer to do things like that after the NNBD migration and do a non-breaking change in this package.

Reply:
We should sync on this plan. I really don't like the idea of handcuffing ourselves in our API decisions because of a bad practice. The SDK made breaking changes in order to improve null safe APIs, I don't see why we shouldn't do the same in packages.

Reply:
The sdk is significantly more hamstrung than we are with regards to versioning. There isn't anything preventing us from doing a non-breaking nnbd release followed by a breaking release.

Most users would be able to support both releases I think as well.

Reply:
This already _is_ breaking, we are already lying if we release it as non-breaking.

Reply:
I would be fine with a deprecation period, either before or after the null safety release.

Reply:
Do you mean its breaking outside of the out of order migration case or is that what you are referring to?

Reply:
If we deprecate it we could do leave this as is, and the new api would have the non-nullable return type with the throw.

Reply:
I mean the out of order migration case.

A deprecation may work for this case since we can replace the API, but won't work in all cases.

Reply:
We can always push additional commits to this branch for breaking changes. At least for this initial PR I would prefer to do a clean migration.